orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java

```diff
@@ -6,260 +6,165 @@
-		}
-		return queryMap;
-	}
-
-	private Map<AssociationKeyMetadata, OrientDBAssociationQueries> initializeAssociationQueries(SessionFactoryImplementor sessionFactoryImplementor) {
-		Map<AssociationKeyMetadata, OrientDBAssociationQueries> queryMap = new HashMap<AssociationKeyMetadata, OrientDBAssociationQueries>();
-		Collection<CollectionPersister> collectionPersisters = sessionFactoryImplementor.getCollectionPersisters().values();
-		for ( CollectionPersister collectionPersister : collectionPersisters ) {
-			if ( collectionPersister instanceof OgmCollectionPersister ) {
-				OgmCollectionPersister ogmCollectionPersister = (OgmCollectionPersister) collectionPersister;
-				EntityKeyMetadata ownerEntityKeyMetadata = ( (OgmEntityPersister) ( ogmCollectionPersister.getOwnerEntityPersister() ) ).getEntityKeyMetadata();
-				AssociationKeyMetadata associationKeyMetadata = ogmCollectionPersister.getAssociationKeyMetadata();
-				queryMap.put( associationKeyMetadata, new OrientDBAssociationQueries( ownerEntityKeyMetadata, associationKeyMetadata ) );
-			}
-		}
-		return queryMap;
-	}
-
-	@Override
-	public GridType overrideType(Type type) {
-		log.info( "overrideType:" + type.getName() + ";" + type.getReturnedClass() );
-		GridType gridType = null;
-		if ( type.getName().equals( "com.orientechnologies.orient.core.id.ORecordId" ) ) {
-			gridType = ORecordIdGridType.INSTANCE;
-		}
-		else {
-			gridType = super.overrideType( type ); // To change body of generated methods, choose Tools | Templates.
-		}
-		return gridType;
-	}
+public class OrientDBDialect extends BaseGridDialect {
+
+    private static final Log log = LoggerFactory.getLogger();
+    private OrientDBDatastoreProvider provider;
+
+    public OrientDBDialect(OrientDBDatastoreProvider provider) {
+        this.provider = provider;
+    }
+
+
+    public ODocument findRecordByKey(ODatabaseDocumentTx executionEngine, EntityKey key, TupleContext tupleContext) {
+        //TODO: include tupleContext selectableColumns projection ??
+        OSQLSynchQuery<ODocument> orientQuery = OrientDBQueryHelper.createSelect(key);
+        return executionEngine.command(orientQuery).execute();
+    }
+
+    private ODocument prepareObjectWithPk(EntityKey key) {
+        ODocument object = new ODocument();
+        String[] columnNames = key.getColumnNames();
+        Object[] columnValues = key.getColumnValues();
+        for (int i = 0; i < columnNames.length; i++) {
+            object.field(columnNames[i], columnValues[i]);
+        }
+        return object;
+    }
+
+    @Override
+    public Tuple getTuple(EntityKey key, TupleContext tupleContext) {
+        ODocument document = findRecordByKey(provider.getConnection(), key, tupleContext);
+        OrientDBTupleSnapshot tuple;
+        if (document == null) {
+            tuple = new OrientDBTupleSnapshot(prepareObjectWithPk(key), key.getMetadata(), OrientDBTupleSnapshot.SnapshotType.INSERT);
+        } else if (isInTheQueue(key, tupleContext)) {
```

Comment:
MongoDB is a `BatchableGridDialect` and uses the queue to keep track of the operations to execute. However, we are not happy about the way we implemented that part and for now I would avoid to use the same approach for OrientDB.